tests/compiler/infer-array.release.wat

```diff
@@ -1646,11 +1646,9 @@
    br_if $folding-inner0
    global.get $~lib/memory/__stack_pointer
    local.tee $0
-   i64.const 0
-   i64.store $0
-   local.get $0
-   i64.const 0
-   i64.store $0 offset=8
+   i32.const 0
+   i32.const 108
+   memory.fill $0
```

Comment:
Hmm, it looks like this signifintally affect to some cases like this one. When we got `memory.fill` instead of two stores

Reply:
This could be a side-effect of the shadow stack now being less compact, so where previously it had two slots it now has more, producing a fill. I also had to increase the default stack size because of it. Would say this is in fact the stack effects are the most notable downside.

Reply:
After increased stacksize this downgrade change still present, hmm

Reply:
Yes, there are more locals now, so shadow stack frames become larger. The cause is that locals are no longer shared, but there is one per named local now.

Reply:
I'm not sure that's a good change. If binaryen could handle all this unique locals into the same number as before (at least for release) well maybe this simplification would make sense, but apparently this simplification will increase optimization time, increase size and degrade speed

Reply:
It's a good change for every other reason, just not for this one, yeah. I'd argue we should never have had shared locals to start with. Merely a relic from the past.

Reply:
```
debug after (this branch)
╭───────────────────────────────╮
│ Stats                         │
╞════════════╤════════════╤═════╡
│ parse      │ 441.393 ms │ 164 │
│ initialize │  73.113 ms │   1 │
│ compile    │ 863.634 ms │   1 │
│ emit       │ 707.080 ms │   4 │
│ validate   │  29.481 ms │   1 │
│ optimize   │   0.262 ms │   1 │
│ transform  │        n/a │   0 │
├────────────┴────────────┴─────┤
│ Took 2259.008 ms              │
│ 85 reads, 5 writes            │
╰───────────────────────────────╯

debug before (main)
╭───────────────────────────────╮
│ Stats                         │
╞════════════╤════════════╤═════╡
│ parse      │ 441.065 ms │ 166 │
│ initialize │  65.055 ms │   1 │
│ compile    │ 848.524 ms │   1 │
│ emit       │ 721.066 ms │   4 │
│ validate   │  29.356 ms │   1 │
│ optimize   │   0.370 ms │   1 │
│ transform  │        n/a │   0 │
├────────────┴────────────┴─────┤
│ Took 2258.351 ms              │
│ 87 reads, 5 writes            │
╰───────────────────────────────╯
```
```
release after (this branch)
╭─────────────────────────────────╮
│ Stats                           │
╞════════════╤══════════════╤═════╡
│ parse      │   233.861 ms │ 164 │
│ initialize │    41.450 ms │   1 │
│ compile    │   546.411 ms │   1 │
│ emit       │   467.417 ms │   4 │
│ validate   │    25.952 ms │   1 │
│ optimize   │ 15297.454 ms │   1 │
│ transform  │          n/a │   0 │
├────────────┴──────────────┴─────┤
│ Took 16661.933 ms               │
│ 85 reads, 4 writes              │
╰─────────────────────────────────╯

release before (main)
╭─────────────────────────────────╮
│ Stats                           │
╞════════════╤══════════════╤═════╡
│ parse      │   231.559 ms │ 166 │
│ initialize │    40.276 ms │   1 │
│ compile    │   553.251 ms │   1 │
│ emit       │   430.730 ms │   4 │
│ validate   │    26.632 ms │   1 │
│ optimize   │ 15514.361 ms │   1 │
│ transform  │          n/a │   0 │
├────────────┴──────────────┴─────┤
│ Took 16842.804 ms               │
│ 87 reads, 4 writes              │
╰─────────────────────────────────╯
```

Reply:
We still have `memory.fill`. Perhaps make sense to increase this value for `frameSize` as well to `24`?
https://github.com/AssemblyScript/assemblyscript/blob/main/src/passes/shadowstack.ts#L276

Reply:
The question there is: How many subsequent stores are about equal to one memory fill memory.fill - i.e. where is the break even point between subsequent stores and a memory.fill? The answer to that question is independent of frame size (only depends on instruction costs), though I'm not sure anymore where 16 comes from.

Reply:
16 - 2 `i64.store`'s
24 - 3 `i64.store`'s
32 - 4 `i64.store`'s

I think 3-4 `i64.store`'s` will be faster than `memory.fill` due to bounds checking

Reply:
I am open to change the threshold, but we should probably measure this somehow by mimicking the actual operations performed on the shadow stack. We could do that in a separate PR, as it is independent of the frame size changes. Wdyt?

Reply:
Ah, wait! Instead of fill 2 stores (16 bytes) we now fill `108` bytes. Hmm

Reply:
On another note, doesn't Binaryen do this already? That is, take a `memory.fill` and decompose it into individual `store`s if that's faster? Perhaps we don't need the optimization in the compiler itself when Binaryen reasons about this already?

Reply:
> That is, take a memory.fill and decompose it into individual stores if that's faster?

Yes. binaryen do this. But `108` is too large for such optimization.

Reply:
> Ah, wait! Instead of fill 2 stores (16 bytes) we now fill `108` bytes. Hmm

In some cases the frame can become significantly larger, yeah. Given other more significant costs, this doesn't appear to impact performance in a noticeable way, though. Having and maintaining a shadow stack in the first place already overshadows costs from individual instructions I guess, and the overall amount of spilling doesn't change, just how large the frame is.

Reply:
I still not sure about all of this. How about do this simplification only when we remove Shadow Stack? That is, when we switch completely to Wasm GC. It seems that it is the shadow stack that will suffer the most after this PR

Reply:
You're share perf data for bootstraping. But as far as I remember it doesn't use incremental GC and therefore shadow stack right?

Reply:
The shadow stack is already quite a suffering on its own (that's why using the minimal runtime has its benefits), and it appears that frame size ultimately doesn't matter much except that the shadow stack becomes larger when there are many scoped locals. On the contrary, this change unblocks quite a few annoying things, like that we now can have actual debug names on all local, and ofc that all this complexity, that like one person on the planet has an idea how to approach, is gone from the compiler.

Reply:
> You're share perf data for bootstraping. But as far as I remember it doesn't use incremental GC and therefore shadow stack right?

It uses incremental GC:

https://github.com/AssemblyScript/assemblyscript/blob/78b2d1afef01995c476b6fcb1647258053a0b193/src/asconfig.json#L10

Reply:
To give an outlook towards what this ultimately enables: Closures need to retroactively lift locals to a captured environment, and the obvious way to do this in an almost-single-pass compiler is to compile a function, find that it closed over a variable, and then amend any function also using that closed over variable. This is easy if each local has a unique index, but extremely hard with local sharing in place. In a sense, we previously performed optimizations that lost critical information, preventing us from creating passes that modify already generated code. No sane compiler would actually do this, I'd say, and would actually do the opposite, like utilize SSA to actually be able to reason about this stuff. Now we don't need SSA for closures, but compacting locals would make this task incredibly hard. Makes a lot of sense to remove this optimization, as it simply happens too early in the pipeline.

Reply:
Now I see a reason. Thanks!

Reply:
Of course you are right that the ideal opportunity to remove this is once we generate Wasm GC code, so Binaryen has all the information, optimizing *before* we'd actually downlevel to a bundled runtime with a shadow stack. Then, Binaryen does the compaction, and we only generate the shadow stack after, yielding about the same result as before. That's the ultimate goal of course, but still we'll have to start somewhen, and the data suggests that this somewhen can as well be now so we can move forward.